### PR TITLE
Add initial codegen support for Cluster object

### LINKED ISSUE #8895: CG1: Update ZAP codegen to generate cluster object definitions
#### Problem
Once #8891 is completed and an example template has been provided for how code-generation for cluster objects should look like, ZAP code-gen should be updated to generate these definitions for all clusters.

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Fix #8895 

#### Change overview
Add initial codegen support for Cluster object

#### Testing
The existing unit test covers this.

### CODE REVIEW

File: src/app/tests/TestDataModelSerialization.cpp [hunk 1]
```diff
@@ -162,7 +162,7 @@ void TestDataModelSerialization::TestDataModelSerialization_EncAndDecSimpleStruc
 
         t.a = 20;
         t.b = true;
-        t.c = clusters::TestCluster::SimpleEnum::VALUEA;
+        t.c = clusters::TestCluster::SimpleEnum::SIMPLE_ENUM_VALUE_A;
```
Comment (mrjerryjohns) [inline src/app/tests/TestDataModelSerialization.cpp:165]: `SimpleEnum` already name spaces the value, so repeating the term `SIMPLE_ENUM` in the enum value is not necessary.

File: zzz_generated/lighting-app/zap-generated/CHIPClusters.cpp [hunk 2]
```diff
@@ -45,6 +45,120 @@ namespace Controller {
...
+exit:
+    // On error, we are responsible to close the sender.
+    if (err != CHIP_NO_ERROR && sender != nullptr)
+    {
+        sender->Shutdown();
+    }
+    return err;
+}
+
+CHIP_ERROR OnOffCluster::On(Callback::Cancelable * onSuccessCallback, Callback::Cancelable * onFailureCallback)
```
Comment (mrjerryjohns) [inline zzz_generated/lighting-app/zap-generated/CHIPClusters.cpp:86]: Are these related changes?

File: zzz_generated/lighting-app/zap-generated/IMClusterCommandHandler.cpp [hunk 3]
```diff
@@ -3223,58 +3223,7 @@ namespace OnOff {
 
 void DispatchServerCommand(CommandHandler * apCommandObj, CommandId aCommandId, EndpointId aEndpointId, TLV::TLVReader & aDataTlv)
 {
-    // We are using TLVUnpackError and TLVError here since both of them can be CHIP_END_OF_TLV
```
Comment (mrjerryjohns) [inline zzz_generated/lighting-app/zap-generated/IMClusterCommandHandler.cpp:3226]: Are these related changes?

**Comment by mrjerryjohns** [comment]
Need to add the BUILD.gn for the cluster-objects library as a dependency to all the existing sample apps.

File: zzz_generated/lighting-app/zap-generated/IMClusterCommandHandler.cpp (hunk 3 above)
Comment (yunhanw-google) [reply zzz_generated/lighting-app/zap-generated/IMClusterCommandHandler.cpp:3226]: not related, removed. thanks

File: zzz_generated/lighting-app/zap-generated/CHIPClusters.cpp (hunk 2 above)
Comment (yunhanw-google) [reply zzz_generated/lighting-app/zap-generated/CHIPClusters.cpp:86]: not related, removed. thanks

File: src/app/tests/TestDataModelSerialization.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/tests/TestDataModelSerialization.cpp:165]: In xml, we have enum https://github.com/project-chip/connectedhomeip/blob/a9ce3207a39de030ddd86e286696c928d5a5d0d4/src/app/zap-templates/zcl/data-model/chip/wifi-network-diagnostics-cluster.xml#L30

without putting the prefix for those enum, it would complain the compilation error.

20_PERCENTER_DIM_UP_IN_0P5_SECONDS_THEN_FADE_TO_OFF_IN_1_SECOND = 0x00,
    ^
../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:1314:5: error: expected identifier
    802__11A  = 0x00,
    ^
../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:1315:5: error: expected identifier
    802__11B  = 0x01,
    ^
../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:1316:5: error: expected identifier
    802__11G  = 0x02,
    ^

**Comment by yunhanw-google** [comment]
> Need to add the BUILD.gn for the cluster-objects library as a dependency to all the existing sample apps.
just add       "${chip_root}/src/app/common:cluster-objects" as dependacy for chip_data_model.gni, all example code has this now.
thanks

File: zzz_generated/app-common/app-common/zap-generated/cluster-objects.h [hunk 4]
```diff
@@ -0,0 +1,1497 @@
...
+    kCatalogVendorIdFieldId = 0,
+    kApplicationIdFieldId   = 1,
+};
+
+struct Type
+{
+public:
+    uint16_t catalogVendorId;
+    Span<const char> applicationId;
+    CHIP_ERROR Encode(TLV::TLVWriter & writer, uint64_t tag) const;
```
Comment (mrjerryjohns) [inline zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:81]: Would be good to visually separate the public members from the encode/decode methods – perhaps, add a 'public:' between them?￼

File: src/app/tests/TestDataModelSerialization.cpp (hunk 1 above)
Comment (mrjerryjohns) [reply src/app/tests/TestDataModelSerialization.cpp:165]: Not worth it to fix that at this stage.

**Review by mrjerryjohns — APPROVED** [review]

File: zzz_generated/app-common/app-common/zap-generated/cluster-objects.h (hunk 4 above)
Comment (yunhanw-google) [reply zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:82]: just add one newline to virtually separate them, thanks

**Review by jmartinez-silabs — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: zzz_generated/app-common/app-common/zap-generated/cluster-objects.h [hunk 5]
```diff
@@ -0,0 +1,1529 @@
...
+#include <app/data-model/Encode.h>
+#include <app/data-model/List.h>
+#include <app/util/basic-types.h>
+
+namespace chip {
+namespace app {
+namespace clusters {
+
+namespace AccountLogin {
+constexpr ClusterId kClusterId = 0x50E;
```
Comment (eve-cxrp) [inline zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:33]: This here is a duplication of information? `zzz_generated/app-common/app-common/zap-generated/ids/Clusters.h` contains this information. Wouldn't it be better to reference this?

File: zzz_generated/app-common/app-common/zap-generated/cluster-objects.h [hunk 6]
```diff
@@ -0,0 +1,1529 @@
...
+namespace ApplicationBasic {
+constexpr ClusterId kClusterId = 0x50D;
+// Enum for ApplicationBasicStatus
+enum class ApplicationBasicStatus : uint8_t
+{
+    APPLICATION_BASIC_STATUS_STOPPED                  = 0x00,
+    APPLICATION_BASIC_STATUS_ACTIVE_VISIBLE_FOCUS     = 0x01,
+    APPLICATION_BASIC_STATUS_ACTIVE_HIDDEN            = 0x02,
+    APPLICATION_BASIC_STATUS_ACTIVE_VISIBLE_NOT_FOCUS = 0x03,
+};
```
Comment (eve-cxrp) [inline zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:56]: This here is a dupplicate of `zzz_generated/app-common/app-common/zap-generated/enums.h` -> `EmberAfApplicationBasicStatus`
